NodeJS Controllers: Update lesson to flow with the previous lesson (#28665)
TheOdinProject/curriculum: nodeJS/express/controllers.md

- `req` - The request object, representing the incoming HTTP request.
- `res` - The response object, representing the HTTP response that will be sent back to the client.

> - `next` - The function that passes the control to the next middleware function in the chain (we'll get to this later). This is optional.

Reviewer: ```suggestion
- `next` - The function that passes control to the next middleware function in the chain (we'll get to this later). This is optional.
```

The naming conventions for these controllers are usually based on the route they will be attached to e.g. `GET` route -> `getSomething`, `POST` route -> `createSomething`, `DELETE` route -> `deleteSomething`, etc. Nonetheless, there is no fixed rule since Express is not opinionated. It will always be based on you or someone else's conventions, and the requirements of the function.

You should already have your application setup in-place from the previous lesson. Previously, we've defined a route for `/authors/:authorId`, but this time we will define a controller for that route and hook it up with sample data. But to make the example produce something, let's first create a mock database with a function that we can use to retrieve an author by their ID.

Create the following file `db.js` at the root of the project, and add the following code:

```javascript
// db.js

const authors = [
  { id: 1, name: "Bryan" },
  { id: 2, name: "Christian" },
  { id: 3, name: "Jason" },
];

const getAuthorById = async (authorId) => {
  return authors.find(author => author.id === authorId);

> };

Author: done with changing most functions

One thing to note is that Express executes middleware functions in the order we define or register them in our application. This means that the sequence in which we define our middleware functions matters, as it determines the order in which they will be invoked during the request-response cycle. So we need to make sure and be aware that our middleware functions are placed in the correct order. As an example, some packages have middleware functions that changes the `Request` object, and as a result, these middleware functions should be placed at the very top of our application in order for us to be able to see their changes in all of your middleware functions below it.

There is also a special type of middleware function that handles errors, which we will discuss shortly.

### Controllers

> As said earlier, controllers are just functions. They also classify as a middleware (at least in the Express world) that are used by route handlers. However, it's important to note that controller and middleware are distinct concepts. Controllers are a key component of the MVC (Model-View-Controller) pattern, a design approach for organizing software. Middleware, on the other hand, is a core feature of Express that allows you to run code, modify requests, or end the cycle at specific points in the request-response cycle. So we are basically using a core functionality of Express to write out patterns of code.

Reviewer: Thoughts?
```suggestion
As said earlier, controllers are just functions. They also classify as a middleware (at least in the Express world) that are used by route handlers. However, it's important to note that controller and middleware are distinct concepts. Controllers are a key component of the MVC (Model-View-Controller) pattern, a design approach for organizing software. Middleware, on the other hand, is a core feature of Express that allows you to run code, modify requests, or end the cycle at specific points in the request-response cycle. So we are using middleware in Express to implement the "Controller" part of the MVC pattern.
```

> A controller comes into play whenever a request hits the server and a route matches the requested HTTP verb and path. The route determines which controller should handle the request based on the defined middleware chain. The appropriate controller then takes over, performing the necessary actions to fulfill the request.. This could involve retrieving data from the model, processing the data, making decisions based on business logic, or updating the model with new data.

Reviewer: ```suggestion
A controller comes into play whenever a request hits the server and a route matches the requested HTTP verb and path. The route determines which controller should handle the request based on the defined middleware chain. The appropriate controller then takes over, performing the necessary actions to fulfill the request. This could involve retrieving data from the model, processing the data, making decisions based on business logic, or updating the model with new data.
```

Once the controller completes its tasks, it passes the processed data to the view. The view then renders this data into a format suitable for sending back to the client. Typically, this would be HTML. Later, when we cover building APIs, we can also send JSON responses like with the APIs that we've previously used e.g. Giphy API.

The naming conventions for these controllers are usually based on the route they will be attached to e.g. `GET` route -> `getSomething`, `POST` route -> `createSomething`, `DELETE` route -> `deleteSomething`, etc. Nonetheless, there is no fixed rule since Express is not opinionated. It will always be based on you or someone else's conventions, and the requirements of the function.

> You should already have your application setup in-place from the previous lesson. Previously, we've defined a route for `/authors/:authorId` to get the author by their ID, let's now define a controller for that route. But to make the example produce something, let's first create a mock database with a function that we can use to retrieve an author by their ID.

Reviewer: Just a heads up, the current Routes lesson doesn't actually tell people to write the library example code - that was just an example of how routers would be written and why they'd be used.

What I'll do is revisit the Routes lesson and touch up that Routers section to actually make people add that stuff to the basic Express app, then this part's language will make a lot more sense.

I'll get this and the Routes lesson touch-up merged together to avoid confusion over discrepancies in lesson language.

Author: Sure I'll wait for that for now, thanks @MaoShizhong

Reviewer: @fcasibu Sorry it's taken so long!

Have a look at #28763 (feel free to comment on that one if you have any thoughts!) and if it looks good, you can reference that stuff here as appropriate. That should help improve the flow between lessons.

Author: No worries, it looks good. I'll probably work on this on friday when I get my free time. Thanks for your work!

Create the following file at the root of the project, and add the following code:

```javascript
// db.js

const authors = [
  { id: 1, name: "Bryan" },
  { id: 2, name: "Christian" },
  { id: 3, name: "Jason" },
];

const getAuthorById = async (authorId) => {
  return authors.find(author => author.id === authorId);
};

module.exports = { getAuthorById };
```

> The file name, content, and location aren't important here, they only serve to illustrate that we're using a database query function to retrieve an author by their ID.

Reviewer: Suggestion - just reiterating the purpose of the code above and that it's not the controller example coming next.

```suggestion
The file name, content, and location aren't important here. This is just a quick mock "database" and query function so we can call it in a controller.
```

> Then for the controller, create the following file:

Reviewer: Just to be more explicit about the structure
```suggestion
Then for the controller, create the following file within a `controllers` folder:
```

```javascript
// controllers/authorController.js

const { getAuthorById } = require("../db");

> const getAuthorById = async (req, res) => {

Reviewer: Going to need to change the name of either this controller or the mock DB query function. Will then need to change all subsequent mentions accordingly.

Author: Sorry, but what do you mean by this?

Author: I just namespaced the db import

  const { authorId } = req.params;

  const author = await getAuthorById(authorId);

  if (!author) {
    res.status(404).send("Author not found");
    return;
  }

>   res.send(author);

Reviewer: Perhaps keep it as a template string? Otherwise the 404 will send a string but the 200 will send JSON.

};

module.exports = { getAuthorById };
```

> In this example, the `getAuthorById` function is a controller that handles a specific action related to retrieving an author by their ID. We'll use this controller by importing it into the file where we define our routes, and using it like this: `router.get("/:authorId", getAuthorById)`. Let's break down what's happening in this controller:

Reviewer: We should probably do 2 things:
- Move the inline code block to a standalone fenced code block
- In that example, refer to the appropriate router file path (`routes/authorsRouter.js`)

Since I named the authors router  `authorsRouter` in the Routes lesson, we should probable change this example to `authorsRouter.get("/:authorId", getAuthorById)`

> 1. The controller extracts the `authorId` from the request parameters (`req.params.authorId`). This assumes we've defined the parameter with a route, such as `/authors/:authorId`.

Reviewer: Suggestion
```suggestion
1. The route path contains a route parameter (`/authors/:authorId`). The controller extracts the `authorId` from `req.params`.
```

1. It then invokes a database query function `getAuthorById` to retrieve the author data based on the `authorId`.

> 1. If the controller doesn't find the author, it sends a response with a 404 status code and the message Author not found, using `res.status(404).send(...)`. It then returns from the controller function to avoid invoking any other logic in the controller, as sending a response doesn't automatically stop the function execution.

Reviewer: ```suggestion
1. If the controller doesn't find the author, it sends a response with a 404 status code and the message `"Author not found"`, using `res.status(404).send(...)`. It then returns from the controller function to avoid invoking any other logic in the controller, as sending a response doesn't automatically stop the function execution.
```

> 1. If the controller finds the author, it sends a response with a 200 status code and the found author object using `res.send(...)`.

Reviewer: Edit the "and the found author object" with the appropriate string from the earlier comment

1. With the string `next('route')` - Will pass control to the next route handler with the same matching path (if there is one). This only works for `app.METHOD` or `router.METHOD`. Potentially, it can also be the same as just calling `next` with no argument.
1. With the string `next('router')` - Will skip all middleware functions attached to the specific router instance and pass control back out of the router instance. Basically, we exit the router and go back to the parent router, e.g. `app` (yes, the Express app is also just a router under the hood).

Out of the four, we will likely only use the first two, unless we have a very specific need that requires the other two.

### Consolidating what we've learned

> The following is the folder structure we'd end up with if you have been coding along:

Reviewer: 2nd person is appropriate here
```suggestion
The following is the folder structure you'd end up with if you had been coding along:
```

```text
express-app/
├─ errors/
│  ├─ CustomNotFoundError.js
├─ controllers/
│  ├─ authorController.js
├─ routes/
│  ├─ authorRouter.js
│  ├─ ... other routers
├─ app.js
├─ db.js
```

> Creating a controller file that contains well-defined responsibilities to handle "Author".

Reviewer: On reflection, I don't actually think we need the file recap from here to line 503. I think it's enough to just share the file structure diagram and that's it. You've already improved the explicitness of codealong instructions earlier in the lesson (inc. the changes for the Routes lesson). I don't think the comments in the code blocks are vital for the lesson, and if you feel any are, you probably should bring them into the main lesson content up above as appropriate.

But IMHO, we can just go straight to "As an exercise, feel free to create more sample data in our `db.js` file and controllers for the other routes we've previously created." after the file structure diagram.

Author: you're right it's very unconventional as well in the curriculum, will remove it
